asst1/CODE/my_scheduler.c

```diff
@@ -111,19 +111,40 @@ void run_maintenance() {
       if (new_prio == thread->priority) { // do nothing
         prev_ptr = ptr;
         ptr = ptr->next;
-      } else { // remove thread from current queue and add it to lower queue
+      } 
+      else { // remove thread from current queue and add it to lower queue
         thread->priority = new_prio;
         insert_ready_q(thread, new_prio);
         if (ptr == ready_q[i]->head) {
           ready_q[i]->head = ptr->next;
-        } else {
+        } 
+	else {
           prev_ptr->next = ptr->next;
         }
         if (ptr == ready_q[i]->tail) {
-          ready_q[i]->tail = NULL;
+          ready_q[i]->tail = prev_ptr;
         }
+	node_t* temp = ptr;
         ptr = ptr->next;
+	free(temp);
+	fCount++;
       }
     }
   }
 }
+
+	
+void free_data() {
+  //  printf("here\n");
+    tcb* thread = get(all_threads, 2);
```

Comment:
what if main thread isn't the last to exit?

Reply:
you’re right, we could set old thread as a global or maybe we can pass it in some how, and yes no stack forgot about that

Reply:
wdym by "old thread as a global"?

Reply:
since we only free a done thread on the next context switch, the very last thread that runs needs to be free’d at exit but we need a way to access it in free_data

Reply:
won't the last thread also run the scheduler and `atexit`?

Reply:
yes we don’t have to free it’s stack if it has one tbh it’s one thread, and yea i was thinking it my give errors but it didn’t in the atexit

Reply:
no sure if i answered your question but even tho last thread runs scheduler previous-done cleans the stack and it does so on the switch after but there’s no switch after the last thread so need to clean at exit if you want to one thread might not need to

Reply:
ok sure, so we need to clean up `prev_done` but not the calling stack/context

Reply:
kk will free prev done within atexit, that is set